pilot/cmd/pilot-agent/status/server.go

```diff
@@ -502,8 +514,51 @@ func negotiateMetricsFormat(contentType string) expfmt.Format {
 	return expfmt.FmtText
 }
 
-func processMetrics(metrics []byte) []byte {
-	return bytes.ReplaceAll(metrics, []byte("\n\n"), []byte("\n"))
+// metricReader used to replace "\n\n" in stream with '\n'.
+type metricReader struct {
+	buf *bufio.Reader
+}
+
+func NewMetricReader(reader io.Reader) *metricReader {
+	return &metricReader{buf: bufio.NewReader(reader)}
+}
+
+// WriteTo io.copy will call this function first
+func (r *metricReader) WriteTo(w io.Writer) (n int64, err error) {
+	var length int
+	var line []byte
+	var isPrefix, isLastLineFinished, isEmptyLine bool
+	newLine := []byte{'\n'}
+
+	for ; err == nil; line, isPrefix, err = r.buf.ReadLine() {
+		// if line length equal reader buffer size, need to add new line
+		isEmptyLine = len(line) == 0 && !isPrefix && !isLastLineFinished
+		isLastLineFinished = isPrefix
+		if isEmptyLine {
+			continue
+		}
+		length, err = w.Write(line)
+		n += int64(length)
+		if err != nil {
+			return
+		}
+		if !isPrefix {
+			length, err = w.Write(newLine)
+			n += int64(length)
+		}
+	}
+	return
+}
+
+func (r *metricReader) Read(p []byte) (n int, err error) {
+	buf := bytes.NewBuffer(p)
+	buf.Reset()
```

Comment:
why reset?

Reply:
Oh 😢 It is a mistake. Because read is never used.So I want to multiplexing writeTo to implement read.

Implementation using bytes.buffer looks like easy. But byte.buffer can only be written from the start of the slice when using an empty slice as the buffer.So I reset it to make sure that slice has no part filled with 0 value.

But bytes.buffer has the side effect of make slice to grow up.I forget that 🤔

Reply:
@howardjohn Here is refactored, Read will be degraded to read all before reading. It make sure io.Reader is implementation.